### PR TITLE
feat: add dynamic sampler for curriculum learning in RL

### PR DESCRIPTION
**Pull request by zpqiu**
# Dynamic Sampler for Curriculum Learning in RL

> [!NOTE]
> I have completed the main code development and validation experiments for the sampler. I haven't yet added the calling logic to RayPPOTrainer or included the corresponding configuration items. If the maintainers believe this feature is useful and suitable for integration into verl, I will continue to refine this PR. 

## Overview

This PR introduces a `DynamicSampler` class that implements curriculum learning based on sample difficulty. The sampler dynamically adjusts the difficulty distribution of training data based on model performance.

Key features:
- Stratified sampling based on sample difficulty levels
- Automatic difficulty adjustment based on model accuracy
- Support for mixed sampling (from current and harder difficulty levels)
- Compatible with StatefulDataLoader for checkpoint and resume

## Implementation Details

`DynamicSampler` extends `torch.utils.data.Sampler` with the following capabilities:
1. Grouping dataset samples by difficulty level
2. Sampling based on current difficulty level and mixing ratio
3. Dynamically adjusting difficulty level based on model performance
4. Supporting state saving and loading for training resumption

## Integration with RLHF Training

In `ray_trainer.py`, we can integrate `DynamicSampler` by:
1. In the `_create_dataloader` method, deciding whether to use `DynamicSampler` based on configuration
2. Calling `update_sampling_policy` at the end of each training step to update the sampling strategy

```python
# ... existing code ...
def _create_dataloader(self):
    # ... existing code ...
    
    # Use dynamic sampler for curriculum learning
    if self.config.data.get('use_dynamic_sampler', False) and self.train_dataset.difficulties is not None:
        sampler = DynamicSampler(
            data_source=np.arange(len(self.train_dataset)),
            difficulties=self.train_dataset.difficulties,
            total_steps=len(self.train_dataset) * self.config.trainer.total_epochs // self.config.data.train_batch_size,
            batch_size=self.config.data.train_batch_size,
            init_level=self.config.data.get('init_level', 0),
            mix_harder_samples=self.config.data.get('mix_harder_samples', 0.2),
            smooth_window=self.config.data.get('smooth_window', 3),
            threshold_for_next_level=self.config.data.get('threshold_for_next_level', 0.75),
            threshold_for_prev_level=self.config.data.get('threshold_for_prev_level', 0.3),
            seed=self.config.data.get('seed', 42)
        )
    else:
        # Use existing random or sequential sampler
        if self.config.data.shuffle:
            # ... existing code ...
        else:
            # ... existing code ...
    
    # ... existing code ...
# ... existing code ...

# Update sampling policy at the end of training loop
# ... existing code ...
# Add at the end of training loop
if isinstance(self.train_dataloader.sampler, DynamicSampler):
    self.train_dataloader.sampler.update_sampling_policy(metrics["critic/score/mean"])
# ... existing code ...
```

## Configuration Example

```yaml
data:
  use_dynamic_sampler: true
  init_level: 0
  mix_harder_samples: 0.2
  smooth_window: 3
  threshold_for_next_level: 0.75
  threshold_for_prev_level: 0.3
```

## Question

Would this dynamic sampling feature be helpful for verl? Hi @vermouth1992 @PeterSH6, could you have a look?

### CODE REVIEW

**Comment by zpqiu** [comment]
I conducted a comparative experiment based on `Qwen2.5-7B`, using the dataset `pe-nlp/ORZ-MATH-57k-Filter-difficulty`. The difficulty of each problem was calculated as `1 - pass_at_n`. 

The training curves and offline evaluation results from the training process are as follows:

## Training Process

The baseline comparison uses random sampling (red line).
截屏2025-03-12 10 59 34

## Offline Evaluation

Each problem generated 8 responses, with Pass@1 calculated using temperature=0.6 and max-length=24k.  
CL: Curriculum Learning with DynamicSampler; RAND: Random Sampling

  | AIME 2024 | MATH 500 | GPQA
-- | -- | -- | --
7B-Base | 4.58 | 59 | 30.38
7B-Instruct | 12.08 | 77.12 | 41.5
CL-80 step | **17.08** | **82** | **38.75**
CL-100 step | 16.25 | 80.5 | 36.13
CL-120 step | 14.17 | 80.25 | 29.75
CL-140 step | 14.17 | 81.87 | 33.75
CL-160 step | 16.67 | 82 | 36
RAND-80 step | 16.25 | 79.37 | 35.75
RAND-100 step | 14.58 | 76.75 | 35.25
RAND-120 step | 13.75 | 79 | 37.38
RAND-140 step | 14.58 | 79.5 | 37

- Curriculum learning achieved peak performance at 80 steps (difficulty levels 1-7).
- Overall, curriculum learning models performed better on MATH500 but showed comparable results to random sampling on AIME.
- Training beyond 80 steps on harder difficulty levels provided no additional benefits, particularly showing no improvement on AIME.

**Comment by zsychina** [comment]
What a great work! Can you provide a quickstart script for hands on test?

**Comment by oleole** [comment]
why do you think curriculum learning does not help AIME performance and steps beyond 80?
